# Confidence intervals case study using NHANES data


This notebook demonstrates how to use Python and its statistical libraries to construct confidence intervals for proportions and means.  We will also cover some important points relating to the properties of confidence intervals, and discuss how to use and interpret confidence intervals in practice.  We will use the 2015-2016 wave of the [NHANES](https://www.cdc.gov/nchs/nhanes/index.htm) data for all the analyses below.

It is important to note that the NHANES data are a "complex survey".  The data are not an independent and representative sample from the target population.  Proper analysis of complex survey data should make use of additional information about the manner in which the data were collected.  Since complex survey analysis is a somewhat specialized topic, we ignore this aspect of the data here, and analyze the NHANES data as if it were an independent and identically distributed sample from a population.

The following module import statements and data reading statement are identical to what we have used previously in course 1.

In [2]:
%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns
import statsmodels.api as sm
import matplotlib.pyplot as plt

In [3]:
df = pd.read_csv('data/nhanes_2015_2016.csv')
df.head(3)

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   

   DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
0       1.0       5.0  ...   124.0    64.0   94.8  184.5    27.8    43.3   
1       2.0       3.0  ...   140.0    88.0   90.4  171.4    30.8    38.0   
2       1.0       3.0  ...   132.0    44.0   83.4  170.1    28.8    35.6   

   BMXARML  BMXARMC  BMXWAIST  HIQ210  
0     43.6     35.9     101.1     2.0  
1     40.0     33.2     107.9     NaN  
2     37.0     31.0     116.5     2.0  

[3 rows x 28 columns]

## Confidence intervals for one proportion

In this section, we demonstrate the construction of confidence intervals for the proportion of people who smoke.  The specific definition of "smoker" used here ([SMQ020](https://wwwn.cdc.gov/Nchs/Nhanes/2015-2016/SMQ_I.htm#SMQ020)) identifies a person as being a smoker if they self-report as having smoked 100 or more cigarettes in their lifetime.  It is more accurate to refer to this as a measure of "lifetime smoking" rather than "current smoking".  Recall that the definitions of these and other NHANES variables can be found using the NHANES code books, or by searching using the link below.

https://wwwn.cdc.gov/nchs/nhanes/search/default.aspx

We will calculate the proportions of smokers separately for females and for males.  Initially we can compare these two proportions and their corresponding confidence intervals informally, but later we will discuss methods to compare two proportions formally using confidence intervals.

First we replace the numeric codes in the variables of interest with text labels, and set the rare answers other than "yes" and "no" to be missing (so they will automatically be omitted from all the analyses below).

In [4]:
df['SMQO20x'] = df['SMQ020'].replace({1: 'yes',
                                      2: 'no',
                                      7: np.nan,
                                      9: np.nan})
df['RIAGENDRx'] = df['RIAGENDR'].replace({1: 'male',
                                          2: 'female'})


We can now [tabulate](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.crosstab.html) the numbers of female and male smokers and non-smokers:

In [18]:
dx = df[['SMQO20x', 'RIAGENDRx']].dropna()
dx.head(3)

SMQO20x RIAGENDRx
0     yes      male
1     yes      male
2     yes      male

In [6]:
pd.crosstab(dx.SMQO20x, dx.RIAGENDRx)

RIAGENDRx  female  male
SMQO20x                
no           2066  1340
yes           906  1413

In [21]:
dx

SMQO20x RIAGENDRx
0        yes      male
1        yes      male
2        yes      male
3         no    female
4         no    female
...      ...       ...
5730     yes    female
5731      no      male
5732     yes    female
5733     yes      male
5734      no    female

[5725 rows x 2 columns]

The confidence interval (CI) is constructed using two inputs: the sample proportion of smokers, and the total sample size for smokers and non-smokers combined. We calculate these values next.

In [25]:
dx['SMQO20x'].replace({'yes': 1, 'no': 0}, inplace=True)
dx['RIAGENDRx'].replace({'male': 1, 'female': 0}, inplace=True)

dz = dx.groupby('RIAGENDRx').agg(['mean', 'count'])
dz.columns = ['Proportion', 'total']
dz

Proportion  total
RIAGENDRx                   
0            0.304845   2972
1            0.513258   2753

Confidence intervals are closely connected to standard errors. Recall that the standard error essentially tells you how far you should expect an estimate to fall from the truth. A confidence interval is an interval that under repeated sampling covers the truth a defined proportion of the time. In most settings, this "coverage probability" is set to 95%.

It turns out that in many settings, a 95% confidence interval can be constructed as the interval consisting of all points that are within two (or 1.96) standard errors of the point estimate. More concisely, the confidence interval approximately spans from e - 2•SE to e + 2•SE, where e is the point estimate and SE is the standard error.

Since the standard error plays such an important role here, we calculate it separately first.

In [28]:
dz.Proportion[0]

0.30484522207267833

In [29]:
# female
p = dz.Proportion[0]
n = dz.total[0]
se_female = np.sqrt(p * (1 - p) / n)

# male
p = dz.Proportion[1]
n = dz.total[1]
se_male = np.sqrt(p * (1 - p) / n)

se_female, se_male

(0.008444152146214435, 0.009526078653689868)

We can see that the standard errors for the estimated proportions of females and males who smoke are similar, and are each around 1% (since we are studying a proportion here, 0.01 corresponds to a 1 percentage point change in the smoking rate).

The standard error for a proportion is maximized when the true proportion is around 1/2, and gets smaller as the true proportion approaches either 0 or 1. The estimated male smoking proportion is closer to 1/2 than the estimated female smoking proportion, and the male sample size is smaller than the female sample size. Both of these factors lead to the male standard error being larger than the female standard error, although the difference is very small in this case.

Next we calculate the 95% confidence intervals for the proportions of female and male smokers using the formula for the one-sample confidence interval for a proportion:

In [31]:
# female
p = dz.Proportion[0]
n = dz.total[0]

moe = 1.96 * np.sqrt(p * (1 - p) / n)
lower_bound, upper_bound = p - moe, p + moe
lower_bound, upper_bound

(0.288294683866098, 0.32139576027925865)

The results above indicate that any population proportion (for female lifetime smokers) between 0.288 and 0.321 would be compatible with the data that we observed in NHANES.

In [32]:
# male
p = dz.Proportion[1]
n = dz.total[1]

moe = 1.96 * np.sqrt(p * (1 - p) / n)
lower_bound, upper_bound = p - moe, p + moe
lower_bound, upper_bound

(0.49458714955108174, 0.531929377873546)

These results indicate that any population proportion (for male lifetime smokers) between 0.493 and 0.531 would be compatible with the NHANES data.

In a routine data analysis, we do not need to calculate these intervals manually. We can use the Statsmodels library to calculate the CI for us in one line:

In [34]:
# male
print('Male:  ', sm.stats.proportion_confint(1413, 1413 + 1340))

# female
print('Female:', sm.stats.proportion_confint(906, 906 + 2066))

Male:   (0.49458749263718593, 0.5319290347874418)
Female: (0.2882949879861214, 0.32139545615923526)


The confidence intervals for the proportions of female and male smokers shown above are quite narrow and do not overlap.  This suggests that there is a substantial difference between the lifetime smoking rates for women and men. However there is no explicit information here about how different the two population proportions might be.  To address this question, we can form a confidence interval for the difference between the proportion of females who smoke and the proportion of males who smoke.

The point estimate of the difference between female and male smoking rates is -0.208 (0.305 - 0.513).  That is, the smoking rate is about 20 percentage points higher in men than in women.  This difference of around 20 percentage points is only a point estimate of the underlying true value -- it is not exactly equal to the difference between the unknown proportions of females and males who smoke in the population.  A confidence interval helps us assess how far the estimated difference may be from the true difference.

As above, we start with the standard error.  The difference between two sample proportions based on independent data has a standard error that reflects the combined uncertainty in the two proportions being differenced.  This standard error can be calculated very easily.  If SE1 and SE2 are the standard errors for two proportions, then sqrt(SE1^2 + SE2^2) is the standard error for the difference of these proportions (`sqrt` is the square root function).  Note that this formula is only accurate if the two sample proportions being differenced are independent.

In the next cell we calculate the standard error for the difference between the proprotion of females who smoke and the proportion of males who smoke.

In [35]:
se_diff = np.sqrt(se_female**2 + se_male**2)
se_diff

0.012729881381407434

The standard error of around 0.013 indicates that the estimated difference statistic -0.208 is expected to fall around 0.013 units from the true value. We do not know in which direction the error lies, and we do not know that the error is exactly 0.013, only that it is around this large on average. For most purposes, a standard error of 0.013 relative to an observed difference of -0.21 would be considered very small. That is, we have a very accurate estimate of the difference between smoking rates in women and in men.

Now that we have the standard error, we can construct a 95% confidence interval for the difference in proportions by taking the estimate and subtracting and adding two (or 1.96) standard errors from it.

In [36]:
d = dz.Proportion[0] - dz.Proportion[1]
lcb = d - 2 * se_diff
ucb = d + 2 * se_diff
print(lcb, ucb)

-0.2338728044024504 -0.18295327887682067


The 95% confidence interval above shows us that any value for the difference of population proportions (between females and males) lying between -0.233 and -0.183 is consistent with the observed data.

## Confidence intervals for sub-populations

Since smoking rates vary strongly with age, it might be more informative to stratify the data into homogeneous age bands and compare the proportions of female and male smokers within each age band. We can also calculate the 95% confidence interval for this difference within each age band. These data can be displayed as a plot, with the difference in proportions plotted as a curve. The confidence intervals can then be used to construct a "confidence band" around the estimates.


In [46]:
# calculate the smoking rates within age/gender groups
df["agegrp"] = pd.cut(df.RIDAGEYR, [18, 30, 40, 50, 60, 70, 80])
pr = df.groupby(["agegrp", "RIAGENDRx"]).agg({"SMQ020": lambda x: np.mean(x == 1)}).unstack()
pr.columns = ["Female", "Male"]
pr

Female      Male
agegrp                      
(18, 30]  0.226601  0.349265
(30, 40]  0.286920  0.502183
(40, 50]  0.268924  0.448878
(50, 60]  0.421277  0.572687
(60, 70]  0.374150  0.654462
(70, 80]  0.324390  0.649254

In [48]:
# The number of people for each calculated proportion
dn = df.groupby(['age_group', 'RIAGENDRx']).agg({'SMQ020': np.size}).unstack()
dn

SMQ020     
RIAGENDRx female male
age_group            
(18, 30]     609  544
(30, 40]     474  458
(40, 50]     502  401
(50, 60]     470  454
(60, 70]     441  437
(70, 80]     410  402